src/test/test-execute.c

```diff
@@ -218,6 +220,38 @@ static void start_parent_slices(Unit *unit) {
         }
 }
 
+static bool can_use_unprivileged_userns(void) {
+        _cleanup_free_ char *v = NULL;
+        int r;
+
+        /* We require kernel.unprivileged_userns_clone=1 to use unprivileged
+         * user namespaces. */
+        r = sysctl_read("kernel/unprivileged_userns_clone", &v);
```

Comment:
hmm, that sysctl doesn't exist here. Is that a downstream ubuntu patch?

wouldn't it make more sense to write a proper check function that just tries to do this, i.e. forks a tiny process off, drops all caps in there, then tries to unshare the userns in there. if that works, it exits immediate with retval 0, and if that doesn't with a privilege error it returns 1, and in case of other errors returns 2. the parent just waits for the result. and proceeds.

that would then work for any kernel, patched or not, and any MAC

Reply:
Do you mean the `kernel.apparmor_restrict_unprivileged_userns` sysctl? I *think* it will become part of the mainline, but currently it is an Ubuntu patch.

In any case, yes, your idea makes more sense. I'll do that instead. Thanks!